packages/loopring_v3/contracts/impl/auxiliaries/DynamicDowntimePriceProvider.sol

```diff
@@ -0,0 +1,126 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity 0.5.10;
+
+import "../../lib/Claimable.sol";
+import "../../lib/MathUint.sol";
+
+import "../../iface/IDowntimePriceProvider.sol";
+
+
+/// @title DynamicDowntimePriceProvider
+/// @author Daniel Wang  - <daniel@loopring.org>
+contract DynamicDowntimePriceProvider is IDowntimePriceProvider, Claimable
+{
+    using MathUint for uint;
+
+    uint public basePrice;
+    uint public maxPenalty;
+    uint public gracePeriodMinutes;
+    uint public gracePeriodPrice;
+    uint public maxNumDowntimeMinutes;
+
+    event SettingsUpdated(
+        uint oldBasePrice,
+        uint oldMaxPenalty,
+        uint oldGracePeriodMinutes,
+        uint oldGracePeriodPrice,
+        uint oldMaxNumDowntimeMinutes
+    );
+
+    constructor(
+        uint _basePrice,
+        uint _maxPenalty,
+        uint _gracePeriodMinutes,
+        uint _gracePeriodPrice,
+        uint _maxNumDowntimeMinutes
+        )
+        Claimable()
+        public
+    {
+        updateSettings(
+            _basePrice,
+            _maxPenalty,
+            _gracePeriodMinutes,
+            _gracePeriodPrice,
+            _maxNumDowntimeMinutes
+        );
+    }
+
+    function getDowntimePrice(
+        uint  totalTimeInMaintenanceSeconds,
+        uint  totalDEXLifeTimeSeconds,
+        uint  numDowntimeMinutes,
+        uint  /* exchangeStakedLRC */,
+        uint  durationToPurchaseMinutes
+        )
+        external
+        view
+        returns (uint)
+    {
+        uint total = numDowntimeMinutes.add(durationToPurchaseMinutes);
```

Comment:
Not sure if I understand this. `numDowntimeMinutes` is the number of downtime minutes the DEX has purchased that are still left (though this value is only updated after calling `stopMaintenanceMode`). If the DEX purchases 120 minutes and calls `stopMaintenanceMode` after 90 minutes then `numDowntimeMinutes == 30`. The DEX only needs to call `startOrContinueMaintenanceMode` immediately after stopping to get the `gracePeriodPrice` cost again which seems like it will be cheaper than the normal price.

Reply:
good point, it is a bug.

Reply:
Fixed, please take another look.